File: services/nuldoc/nuldoc-src/components/SlidePageEntry.ts
import {
  getPostPublishedDate,
  getPostUpdatedDate,
  postHasAnyUpdates,
} from "../generators/post.ts";
import { SlidePage } from "../generators/slide.ts";
import { dateToString } from "../revision.ts";
import { Config } from "../config.ts";
import {
  a,
  article,
  elem,
  Element,
  footer,
  h2,
  header,
  p,
  section,
} from "../dom.ts";
import TagList from "./TagList.ts";

type Props = { slide: SlidePage; config: Config };

export default function SlidePageEntry({ slide, config }: Props): Element {
  return article(
    { class: "post-entry" },
    a(
      { href: slide.href },
      header(
        { class: "entry-header" },
        h2({}, slide.description),
      ),
      section({ class: "entry-content" }, p({}, slide.title)),
      footer(
        { class: "entry-footer" },
        elem(
          "time",
          { datetime: dateToString(getPostPublishedDate(slide)) },
          dateToString(getPostPublishedDate(slide)),
        ),
        " 登壇",
        postHasAnyUpdates(slide) ? "、" : null,
        postHasAnyUpdates(slide)
          ? elem(
            "time",
            { datetime: dateToString(getPostUpdatedDate(slide)) },
            dateToString(getPostUpdatedDate(slide)),
          )
          : null,
        postHasAnyUpdates(slide) ? " 更新" : null,
        slide.tags.length !== 0 ? TagList({ tags: slide.tags, config }) : null,
      ),
    ),
  );
}

File: services/nuldoc/lib/nuldoc/components/slide_page_entry.rb
module Nuldoc
  module Components
    class SlidePageEntry < DOM::HTMLBuilder
      def initialize(slide:, config:)
        super()
        @slide = slide
        @config = config
      end

      def build
        published = Revision.date_to_string(GeneratorUtils.published_date(@slide))
        updated = Revision.date_to_string(GeneratorUtils.updated_date(@slide))
        has_updates = GeneratorUtils.any_updates?(@slide)

        article class: 'post-entry' do
          a href: @slide.href do
            header(class: 'entry-header') { h2 { text @slide.title } }
            section(class: 'entry-content') { p { text @slide.description } }
            footer class: 'entry-footer' do
              time(datetime: published) { text published }
              text ' 登壇'
              if has_updates
                text '、'
                time(datetime: updated) { text updated }
                text ' 更新'
              end
              render TagList, tags: @slide.tags, config: @config if @slide.tags.length.positive?
            end
          end
        end
      end
    end
  end
end
